### PR TITLE
feat(SINAN): moving EGH changes to PySUS

### PR DESCRIPTION
**Pull request by luabida**
related to https://github.com/thegraphnetwork/epigraphhub_py/pull/211
related to https://github.com/thegraphnetwork/EpiGraphHub/pull/170

### CODE REVIEW

**Comment by luabida** [comment]
Working in progress

**Comment by luabida** [comment]
@fccoelho ready for review

File: pysus/classes/sinan/__init__.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+import pyarrow.parquet as pq
+from dbfread import DBF
+from pysus.utilities.readdbc import dbc2dbf
+from sqlalchemy import DATE, INTEGER, NUMERIC, VARCHAR
+
+from .diseases import DISEASE_CODE
+from .typecast import COLUMN_TYPE
+
+
+class SINAN:
```
Comment (fccoelho) [inline pysus/classes/sinan/__init__.py:18]: @luabida if this refactor is going to break the PySUS current API, it must be discussed in more details first.

**Review by fccoelho — COMMENTED** [review]
One style comment: It's not a good Idea to name a package `classes`.
Besides, this is not the pythonic way of structuring the namespace of a package

**Review by fccoelho — CHANGES_REQUESTED** [review]
Please add docstrings to all functions

File: pysus/online_data/__init__.py [hunk 2]
```diff
@@ -240,3 +250,371 @@ def parse(line):
...
+
+    def __checkdatabase__(self, database):
+        if database not in self.available_dbs:
+            raise ValueError(
+                f"{database} not found"
+                f" available databases: {self.available_dbs}"
+            )
+        return database
+
+    def list_all_dbcs(
```
[the remark this answers is not in the document]
Comment (luabida) [reply pysus/online_data/__init__.py:274]: Yes, it also needs refactor on every dataset file to use the new centralized classes

File: pysus/online_data/__init__.py [hunk 3]
```diff
@@ -240,3 +250,371 @@ def parse(line):
...
+            )
+            return str(filepath)
+        except Exception as e:
+            logging.error(f"Not able to download {filename}")
+            raise e
+        finally:
+            ftp.close()
+
+
+class SINAN_Disease:
```
[the remark this answers is not in the document]
Comment (fccoelho) [reply pysus/online_data/__init__.py:531]: I noticed you removed the caching, @luabida, users of PySUS are used to have this local cache to speed up their analyses, if we decide to remove we need to explain in the release notes that caching is no longer available and that users should remember to make local copies of the files they want to access multiple times without having to redownload them.

**Comment by luabida** [comment]
Needs tests, implementation and notebook. I've made some major improvements in the code and some debugging, looks very robust now. I'll only take a close look into how to convert the types now that I have one centralized method that extracts the parquets. Tomorrow morning I've a light surgery and hopefully will finish this before sunday.
Some tests can be done by now like:
```
FTP_Inspect('PNI').list_available_years(UF='RJ')
```
```
FTP_Downloader('CIHA').download(UFs=['AC', 'RJ'], years=[1,2,3], months=[1,2,3])
```
```
FTP_Downloader('SINAN').download(SINAN_disease='Zika', years=[17, 18])
```

File: pysus/online_data/__init__.py [hunk 4]
```diff
@@ -68,6 +68,12 @@ def cache_contents():
 def parquets_to_dataframe(
     parquet_dir: str(PosixPath), clean_after_read=False
 ) -> pd.DataFrame:
+    """ 
+    Receives a parquet directory path and returns it as a
+    dataframe, trying to clean white spaces and convert to
+    the properly data types. Can read only one parquet dir
```
Comment (fccoelho) [inline pysus/online_data/__init__.py:74]: replace 'properly' by 'correct'

Comment (luabida) [reply pysus/online_data/__init__.py]: About this function, the only dtype that objects converts to is the String type, some white spaces in the values can be also found, e.g. `1020         `, that I think are preventing pandas to converts  to the correct dtypes

**Comment by luabida** [comment]
Changes: Now each DBC or DBF will be converted to parquet when downloading. The download can be done using multiple UFs, months and years, and will return a list with downloaded parquets. To read the parquets, there is a function in the `__init__`  file that tries to parse the parquets into dataframe, with the correct dtypes and also removes values that are only blank spaces. I've made a jupyter book about the FTP classes, but it is pretty repetitive, since each DB has the same methods: `lit_available_years`, `list_all` and `download`

**Comment by fccoelho** [comment]
> 

When you say "remove values", what exactly are you doing?  Can you link the line here?

**Comment by luabida** [comment]
It is failing because of these encoding error:

```
geocodigo = '280050   '

    @np.vectorize
    def add_dv(geocodigo):
        if len(str(geocodigo)) == 7:
            return geocodigo
        else:
>           return int(str(geocodigo) + str(calculate_digit(geocodigo)))
E           ValueError: invalid literal for int() with base 10: '280050   6'
```

**Comment by luabida** [comment]
> > 
> 
> When you say "remove values", what exactly are you doing? Can you link the line here?

[here](https://github.com/AlertaDengue/PySUS/blob/a9be440e000401eaf574ceb796481e20feddb53e/pysus/online_data/__init__.py#L86) it will remove every value that is only blank spaces

**Comment by luabida** [comment]
Methods that could be added: `list_available_states`, `list_available_months`

File: pysus/online_data/__init__.py (hunk 4 above)
Comment (fccoelho) [reply pysus/online_data/__init__.py]: After converting an object to string, apply a `strip` to it before trying the next conversion:
```python
d = "1020      "
d_converted = int(str(s).strip())
```

**Comment by fccoelho** [comment]
> 

This is a good example where the `geocodigo` string should be stripped even before passing it to the `add_dv` function. Otherwise it will fail the first `if`.

**Comment by luabida** [comment]
> > 
> 
> This is a good example where the `geocodigo` string should be stripped even before being passed to the `add_dv` function. Otherwise it will fail the first `if`.

SINAN, for instance, some diseases have more than a hundred columns in the dataframe. I can add `geocodigo`, but there will be a lot more like this case

**Comment by luabida** [comment]
@fccoelho do you have some ideas how to fix the failing tests, one of them is because the values are 1-2 and not Masc-Fem, but the other one I'm not so sure. Besides that, the PR is ready to review

**Pull request merged**